File: cli/internal/wikis/template_selection.go
package wikis

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"

	"harness-data/cli/internal/harness"
)

const (
	TemplateSelectionLogicalPath       = "reports/selection.yaml"
	LegacyTemplateSelectionLogicalPath = "templates/selection.yaml"
)

type TemplateDoctorResult struct {
	Status            string                  `json:"status"`
	SelectionPath     string                  `json:"selectionPath"`
	Rules             []TemplateSelectionRule `json:"rules,omitempty"`
	Errors            []string                `json:"errors,omitempty"`
	Warnings          []string                `json:"warnings,omitempty"`
	Suggestions       []TemplateSelectionRule `json:"suggestions,omitempty"`
	SuggestionPath    string                  `json:"suggestionPath,omitempty"`
	SuggestionWritten bool                    `json:"suggestionWritten"`
}

func LoadTemplateSelectionPolicy(root string) (TemplateSelectionPolicy, string, error) {
	resolver, err := harness.NewPathResolver(root)
	if err != nil {
		return TemplateSelectionPolicy{}, "", err
	}
	selectionPath := resolver.Resolve(TemplateSelectionLogicalPath)
	data, err := os.ReadFile(selectionPath)
	if err != nil {
		if os.IsNotExist(err) {
			selectionPath = resolver.Resolve(LegacyTemplateSelectionLogicalPath)
			data, err = os.ReadFile(selectionPath)
			if err != nil {
				if os.IsNotExist(err) {
					return TemplateSelectionPolicy{Version: 1}, resolver.Resolve(TemplateSelectionLogicalPath), nil
				}
				return TemplateSelectionPolicy{}, selectionPath, err
			}
		} else {
			return TemplateSelectionPolicy{}, selectionPath, err
		}
	}
	policy, err := ParseTemplateSelectionYAML(data)
	if err != nil {
		return TemplateSelectionPolicy{}, selectionPath, err
	}
	if policy.Version == 0 {
		policy.Version = 1
	}
	return policy, selectionPath, nil
}

func ParseTemplateSelectionYAML(data []byte) (TemplateSelectionPolicy, error) {
	var policy TemplateSelectionPolicy
	var current *TemplateSelectionRule
	var arrayField string
	lines := strings.Split(string(data), "\n")
	for lineNo, raw := range lines {
		line := strings.TrimRight(raw, " \t\r")
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}
		if !strings.HasPrefix(line, " ") && strings.Contains(trimmed, ":") {
			key, value := splitYAMLKeyValue(trimmed)
			switch key {
			case "version":
				if value == "" {
					return TemplateSelectionPolicy{}, fmt.Errorf("selection.yaml:%d version must be scalar", lineNo+1)
				}
				version, err := strconv.Atoi(cleanYAMLScalar(value))
				if err != nil {
					return TemplateSelectionPolicy{}, fmt.Errorf("selection.yaml:%d invalid version", lineNo+1)
				}
				policy.Version = version
			case "templates":
				if value != "" {
					return TemplateSelectionPolicy{}, fmt.Errorf("selection.yaml:%d templates must be a list", lineNo+1)
				}
			default:
				return TemplateSelectionPolicy{}, fmt.Errorf("selection.yaml:%d unsupported key: %s", lineNo+1, key)
			}
			arrayField = ""
			continue
		}
		if strings.HasPrefix(trimmed, "- ") {
			content := strings.TrimSpace(strings.TrimPrefix(trimmed, "- "))
			if current != nil && arrayField != "" && !strings.Contains(content, ":") {
				appendTemplateSelectionArray(current, arrayField, cleanYAMLScalar(content))
				continue
			}
			policy.Templates = append(policy.Templates, TemplateSelectionRule{})
			current = &policy.Templates[len(policy.Templates)-1]
			arrayField = ""
			if content == "" {
				continue
			}
			key, value := splitYAMLKeyValue(content)
			if key == "" {
				return TemplateSelectionPolicy{}, fmt.Errorf("selection.yaml:%d invalid template item", lineNo+1)
			}
			if err := setTemplateSelectionScalar(current, key, value, lineNo+1); err != nil {
				return TemplateSelectionPolicy{}, err
			}
			continue
		}
		if current == nil {
			return TemplateSelectionPolicy{}, fmt.Errorf("selection.yaml:%d field outside template item", lineNo+1)
		}
		key, value := splitYAMLKeyValue(trimmed)
		if key == "" {
			return TemplateSelectionPolicy{}, fmt.Errorf("selection.yaml:%d invalid field", lineNo+1)
		}
		switch key {
		case "covers", "intents":
			if value == "" {
				arrayField = key
				continue
			}
			values := parseInlineYAMLArray(value)
			for _, item := range values {
				appendTemplateSelectionArray(current, key, item)
			}
			arrayField = ""
		default:
			if err := setTemplateSelectionScalar(current, key, value, lineNo+1); err != nil {
				return TemplateSelectionPolicy{}, err
			}
			arrayField = ""
		}
	}
	return policy, nil
}

func splitYAMLKeyValue(line string) (string, string) {
	parts := strings.SplitN(line, ":", 2)
	if len(parts) != 2 {
		return "", ""
	}
	return strings.TrimSpace(parts[0]), strings.TrimSpace(parts[1])
}

func setTemplateSelectionScalar(rule *TemplateSelectionRule, key, value string, lineNo int) error {
	value = cleanYAMLScalar(value)
	switch key {
	case "id":
		rule.ID = value
	case "playbook":
		rule.Playbook = value
	case "template":
		rule.Template = value
	case "type":
		rule.Type = value
	case "domain":
		rule.Domain = value
	case "priority":
		if value == "" {
			return fmt.Errorf("selection.yaml:%d priority must be numeric", lineNo)
		}
		priority, err := strconv.Atoi(value)
		if err != nil {
			return fmt.Errorf("selection.yaml:%d invalid priority", lineNo)
		}
		rule.Priority = priority
	default:
		return fmt.Errorf("selection.yaml:%d unsupported template field: %s", lineNo, key)
	}
	return nil
}

func appendTemplateSelectionArray(rule *TemplateSelectionRule, key, value string) {
	if value == "" {
		return
	}
	switch key {
	case "covers":
		rule.Covers = append(rule.Covers, value)
	case "intents":
		rule.Intents = append(rule.Intents, value)
	}
}

func ValidateTemplateSelectionPolicy(root string, policy TemplateSelectionPolicy) []string {
	resolver, err := harness.NewPathResolver(root)
	if err != nil {
		return []string{"path config error: " + err.Error()}
	}
	var errs []string
	ids := map[string]bool{}
	for i, rule := range policy.Templates {
		prefix := fmt.Sprintf("templates[%d]", i)
		if rule.ID == "" {
			errs = append(errs, prefix+": id is required")
		} else if ids[rule.ID] {
			errs = append(errs, prefix+": duplicate id "+rule.ID)
		}
		ids[rule.ID] = true
		if rule.Playbook == "" || inferKind(rule.Playbook) != KindPlaybook {
			errs = append(errs, prefix+": playbook must reference a playbook logical path")
		} else if !fileExists(resolver.Resolve(rule.Playbook)) {
			errs = append(errs, prefix+": missing playbook "+rule.Playbook)
		}
		if rule.Template == "" || !isAllowedTemplateSelectionPath(rule.Template) {
			errs = append(errs, prefix+": template must use templates/... or reports/.../template.md")
		} else if !fileExists(resolver.Resolve(rule.Template)) {
			errs = append(errs, prefix+": missing template "+rule.Template)
		}
		if rule.Type != "report" && rule.Type != "composite" && rule.Type != "single" {
			errs = append(errs, prefix+": type must be report, composite, or single")
		}
		for _, cover := range rule.Covers {
			if !isSpecDocPath(cover) {
				errs = append(errs, prefix+": cover must reference a spec logical path: "+cover)
			} else if !fileExists(resolver.Resolve(cover)) {
				errs = append(errs, prefix+": missing cover "+cover)
			}
		}
		if (rule.Type == "report" || rule.Type == "composite") && len(rule.Covers) == 0 && len(rule.Intents) == 0 {
			errs = append(errs, prefix+": report/composite must define covers or intents")
		}
	}
	return errs
}

func BuildTemplateDoctor(root, out string) (TemplateDoctorResult, error) {
	policy, selectionPath, err := LoadTemplateSelectionPolicy(root)
	if err != nil {
		return TemplateDoctorResult{}, err
	}
	errs := ValidateTemplateSelectionPolicy(root, policy)
	suggestions, suggestErr := SuggestTemplateSelection(root, policy)
	if suggestErr != nil {
		errs = append(errs, suggestErr.Error())
	}
	result := TemplateDoctorResult{
		Status:        "PASS",
		SelectionPath: filepath.ToSlash(selectionPath),
		Rules:         policy.Templates,
		Errors:        errs,
		Suggestions:   suggestions,
	}
	if len(errs) > 0 {
		result.Status = "FAIL"
	} else if len(suggestions) > 0 {
		result.Status = "WARN"
		result.Warnings = append(result.Warnings, "selection.yaml is missing high-confidence report/composite templates")
	}
	if len(suggestions) > 0 {
		if out == "" {
			out = filepath.Join(root, "selection.suggested.yaml")
		}
		result.SuggestionPath = filepath.ToSlash(out)
		if err := os.WriteFile(out, []byte(RenderTemplateSelectionYAML(TemplateSelectionPolicy{Version: 1, Templates: suggestions})), 0o644); err != nil {
			return TemplateDoctorResult{}, err
		}
		result.SuggestionWritten = true
	}
	return result, nil
}

func SuggestTemplateSelection(root string, policy TemplateSelectionPolicy) ([]TemplateSelectionRule, error) {
	corpus, _, err := LoadCorpus(root)
	if err != nil {
		return nil, err
	}
	knownTemplates := map[string]bool{}
	for _, rule := range policy.Templates {
		knownTemplates[rule.Template] = true
	}
	var suggestions []TemplateSelectionRule
	for _, doc := range corpus.Docs {
		if doc.Kind != KindTemplate || doc.IsIndex || knownTemplates[doc.Path] {
			continue
		}
		base := pathBase(doc.Path)
		if !isAllowedTemplateSelectionPath(doc.Path) {
			continue
		}
		if strings.HasPrefix(doc.Path, "templates/") && !strings.HasPrefix(base, "r-") && !strings.HasPrefix(base, "c-") {
			continue
		}
		playbook := SamePath(doc.Path, "playbooks")
		spec := SamePath(doc.Path, "spec")
		if corpus.ByPath[playbook] == nil || corpus.ByPath[spec] == nil {
			continue
		}
		ruleType := "report"
		if strings.HasPrefix(base, "c-") {
			ruleType = "composite"
		}
		suggestions = append(suggestions, TemplateSelectionRule{
			ID:       stableTemplateSelectionID(doc.Path),
			Playbook: playbook,
			Template: doc.Path,
			Type:     ruleType,
			Domain:   doc.Domain,
			Covers:   []string{spec},
			Intents:  []string{"report", "diagnosis"},
			Priority: 100,
		})
	}
	sort.Slice(suggestions, func(i, j int) bool { return suggestions[i].ID < suggestions[j].ID })
	return suggestions, nil
}

func RenderTemplateSelectionYAML(policy TemplateSelectionPolicy) string {
	var b strings.Builder
	version := policy.Version
	if version == 0 {
		version = 1
	}
	fmt.Fprintf(&b, "version: %d\n\n", version)
	b.WriteString("templates:\n")
	for _, rule := range policy.Templates {
		fmt.Fprintf(&b, "  - id: %s\n", rule.ID)
		fmt.Fprintf(&b, "    playbook: %s\n", rule.Playbook)
		fmt.Fprintf(&b, "    template: %s\n", rule.Template)
		fmt.Fprintf(&b, "    type: %s\n", rule.Type)
		if rule.Domain != "" {
			fmt.Fprintf(&b, "    domain: %s\n", rule.Domain)
		}
		writeSimpleYAMLArray(&b, "covers", rule.Covers)
		writeSimpleYAMLArray(&b, "intents", rule.Intents)
		fmt.Fprintf(&b, "    priority: %d\n", rule.Priority)
	}
	return b.String()
}

func writeSimpleYAMLArray(b *strings.Builder, key string, values []string) {
	if len(values) == 0 {
		return
	}
	fmt.Fprintf(b, "    %s:\n", key)
	for _, value := range values {
		fmt.Fprintf(b, "      - %s\n", value)
	}
}

func fileExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}

func pathBase(logical string) string {
	parts := strings.Split(logical, "/")
	if len(parts) == 0 {
		return logical
	}
	return parts[len(parts)-1]
}

func stableTemplateSelectionID(templatePath string) string {
	id := strings.TrimSuffix(strings.TrimPrefix(templatePath, "templates/"), ".md")
	id = strings.TrimPrefix(id, "reports/")
	id = strings.ReplaceAll(id, "/", "_")
	id = strings.ReplaceAll(id, "-", "_")
	return id
}

func isReportTemplatePath(logical string) bool {
	return strings.HasPrefix(logical, "reports/") && pathBase(logical) == "template.md"
}

func isAllowedTemplateSelectionPath(logical string) bool {
	return strings.HasPrefix(logical, "templates/") || isReportTemplatePath(logical)
}

File: packages/data-harness-cli/src/lib/wikis/template-selection.js
import { readFileSync, writeFileSync } from "node:fs";
import path from "node:path";

import { newPathResolver } from "../harness.js";
import { isRegularFile, toSlash } from "../fs-utils.js";
import { cleanYAMLScalar, parseInlineYAMLArray, splitYAMLKV } from "./aliases.js";
import { inferKind, isSpecDocPath, loadCorpus } from "./parse.js";
import { KIND_PLAYBOOK, KIND_TEMPLATE, samePath } from "./paths.js";

export const TEMPLATE_SELECTION_LOGICAL_PATH = "reports/selection.yaml";
export const LEGACY_TEMPLATE_SELECTION_LOGICAL_PATH = "templates/selection.yaml";

export function loadTemplateSelectionPolicy(root) {
  const resolver = newPathResolver(root);
  let selectionPath = resolver.resolve(TEMPLATE_SELECTION_LOGICAL_PATH);
  let data;
  try {
    data = readFileSync(selectionPath);
  } catch (error) {
    if (!error || error.code !== "ENOENT") {
      const wrap = new Error(error.message || String(error));
      wrap.cause = error;
      wrap.selectionPath = selectionPath;
      throw wrap;
    }
    selectionPath = resolver.resolve(LEGACY_TEMPLATE_SELECTION_LOGICAL_PATH);
    try {
      data = readFileSync(selectionPath);
    } catch (legacyError) {
      if (!legacyError || legacyError.code !== "ENOENT") {
        const wrap = new Error(legacyError.message || String(legacyError));
        wrap.cause = legacyError;
        wrap.selectionPath = selectionPath;
        throw wrap;
      }
      return {
        policy: { version: 1, templates: [] },
        selectionPath: resolver.resolve(TEMPLATE_SELECTION_LOGICAL_PATH),
      };
    }
  }
  const policy = parseTemplateSelectionYAML(data);
  if (policy.version === 0) policy.version = 1;
  return { policy, selectionPath };
}

export function parseTemplateSelectionYAML(data) {
  const policy = { version: 0, templates: [] };
  let current = null;
  let arrayField = "";
  const text = Buffer.isBuffer(data) ? data.toString("utf8") : String(data ?? "");
  const lines = text.split("\n");
  for (let i = 0; i < lines.length; i++) {
    const lineNo = i + 1;
    const line = lines[i].replace(/[ \t\r]+$/g, "");
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    if (!line.startsWith(" ") && trimmed.includes(":")) {
      const { key, value } = splitYAMLKeyValue(trimmed);
      switch (key) {
        case "version": {
          if (value === "") throw new Error(`selection.yaml:${lineNo} version must be scalar`);
          const version = parseYAMLInt(cleanYAMLScalar(value));
          if (version == null) throw new Error(`selection.yaml:${lineNo} invalid version`);
          policy.version = version;
          break;
        }
        case "templates":
          if (value !== "") throw new Error(`selection.yaml:${lineNo} templates must be a list`);
          break;
        default:
          throw new Error(`selection.yaml:${lineNo} unsupported key: ${key}`);
      }
      arrayField = "";
      continue;
    }
    if (trimmed.startsWith("- ")) {
      const content = trimmed.slice(2).trim();
      if (current && arrayField && !content.includes(":")) {
        appendTemplateSelectionArray(current, arrayField, cleanYAMLScalar(content));
        continue;
      }
      current = emptyTemplateSelectionRule();
      policy.templates.push(current);
      arrayField = "";
      if (content === "") continue;
      const { key, value } = splitYAMLKeyValue(content);
      if (!key) throw new Error(`selection.yaml:${lineNo} invalid template item`);
      setTemplateSelectionScalar(current, key, value, lineNo);
      continue;
    }
    if (!current) throw new Error(`selection.yaml:${lineNo} field outside template item`);
    const { key, value } = splitYAMLKeyValue(trimmed);
    if (!key) throw new Error(`selection.yaml:${lineNo} invalid field`);
    switch (key) {
      case "covers":
      case "intents":
        if (value === "") {
          arrayField = key;
          continue;
        }
        for (const item of parseInlineYAMLArray(value)) {
          appendTemplateSelectionArray(current, key, item);
        }
        arrayField = "";
        break;
      default:
        setTemplateSelectionScalar(current, key, value, lineNo);
        arrayField = "";
    }
  }
  return policy;
}

function splitYAMLKeyValue(line) {
  const kv = splitYAMLKV(line);
  if (!kv) return { key: "", value: "" };
  return kv;
}

function emptyTemplateSelectionRule() {
  return {
    id: "",
    playbook: "",
    template: "",
    type: "",
    domain: "",
    covers: [],
    intents: [],
    priority: 0,
  };
}

function setTemplateSelectionScalar(rule, key, value, lineNo) {
  value = cleanYAMLScalar(value);
  switch (key) {
    case "id":
      rule.id = value;
      return;
    case "playbook":
      rule.playbook = value;
      return;
    case "template":
      rule.template = value;
      return;
    case "type":
      rule.type = value;
      return;
    case "domain":
      rule.domain = value;
      return;
    case "priority": {
      if (value === "") throw new Error(`selection.yaml:${lineNo} priority must be numeric`);
      const priority = parseYAMLInt(value);
      if (priority == null) throw new Error(`selection.yaml:${lineNo} invalid priority`);
      rule.priority = priority;
      return;
    }
    default:
      throw new Error(`selection.yaml:${lineNo} unsupported template field: ${key}`);
  }
}

function appendTemplateSelectionArray(rule, key, value) {
  if (!value) return;
  if (key === "covers") rule.covers.push(value);
  if (key === "intents") rule.intents.push(value);
}

function parseYAMLInt(value) {
  if (!/^[+-]?\d+$/.test(String(value))) return null;
  return Number.parseInt(value, 10);
}

export function validateTemplateSelectionPolicy(root, policy) {
  let resolver;
  try {
    resolver = newPathResolver(root);
  } catch (error) {
    return [`path config error: ${error.message || error}`];
  }
  const errs = [];
  const ids = new Set();
  (policy.templates || []).forEach((rule, i) => {
    const prefix = `templates[${i}]`;
    if (!rule.id) errs.push(`${prefix}: id is required`);
    else if (ids.has(rule.id)) errs.push(`${prefix}: duplicate id ${rule.id}`);
    ids.add(rule.id);
    if (!rule.playbook || inferKind(rule.playbook) !== KIND_PLAYBOOK) {
      errs.push(`${prefix}: playbook must reference a playbook logical path`);
    } else if (!fileExists(resolver.resolve(rule.playbook))) {
      errs.push(`${prefix}: missing playbook ${rule.playbook}`);
    }
    if (!rule.template || !isAllowedTemplateSelectionPath(rule.template)) {
      errs.push(`${prefix}: template must use templates/... or reports/.../template.md`);
    } else if (!fileExists(resolver.resolve(rule.template))) {
      errs.push(`${prefix}: missing template ${rule.template}`);
    }
    if (rule.type !== "report" && rule.type !== "composite" && rule.type !== "single") {
      errs.push(`${prefix}: type must be report, composite, or single`);
    }
    for (const cover of rule.covers || []) {
      if (!isSpecDocPath(cover)) {
        errs.push(`${prefix}: cover must reference a spec logical path: ${cover}`);
      } else if (!fileExists(resolver.resolve(cover))) {
        errs.push(`${prefix}: missing cover ${cover}`);
      }
    }
    if ((rule.type === "report" || rule.type === "composite") && !(rule.covers || []).length && !(rule.intents || []).length) {
      errs.push(`${prefix}: report/composite must define covers or intents`);
    }
  });
  return errs;
}

export function buildTemplateDoctor(root, out = "") {
  const { policy, selectionPath } = loadTemplateSelectionPolicy(root);
  const errors = validateTemplateSelectionPolicy(root, policy);
  let suggestions = [];
  try {
    suggestions = suggestTemplateSelection(root, policy);
  } catch (error) {
    errors.push(error.message || String(error));
  }
  const result = {
    status: "PASS",
    selectionPath: toSlash(selectionPath),
    rules: policy.templates,
    errors,
    warnings: [],
    suggestions,
    suggestionPath: "",
    suggestionWritten: false,
  };
  if (errors.length > 0) result.status = "FAIL";
  else if (suggestions.length > 0) {
    result.status = "WARN";
    result.warnings.push("selection.yaml is missing high-confidence report/composite templates");
  }
  if (suggestions.length > 0) {
    if (!out) out = path.join(root, "selection.suggested.yaml");
    result.suggestionPath = toSlash(out);
    writeFileSync(out, renderTemplateSelectionYAML({ version: 1, templates: suggestions }), { mode: 0o644 });
    result.suggestionWritten = true;
  }
  return result;
}

export function suggestTemplateSelection(root, policy) {
  const { corpus } = loadCorpus(root);
  const knownTemplates = new Set((policy.templates || []).map((rule) => rule.template).filter(Boolean));
  const suggestions = [];
  for (const doc of corpus.docs) {
    if (doc.kind !== KIND_TEMPLATE || doc.isIndex || knownTemplates.has(doc.path)) continue;
    const base = pathBase(doc.path);
    if (!isAllowedTemplateSelectionPath(doc.path)) continue;
    if (doc.path.startsWith("templates/") && !base.startsWith("r-") && !base.startsWith("c-")) continue;
    const playbook = samePath(doc.path, "playbooks");
    const spec = samePath(doc.path, "spec");
    if (!corpus.byPath[playbook] || !corpus.byPath[spec]) continue;
    let ruleType = "report";
    if (base.startsWith("c-")) ruleType = "composite";
    suggestions.push({
      id: stableTemplateSelectionID(doc.path),
      playbook,
      template: doc.path,
      type: ruleType,
      domain: doc.domain || "",
      covers: [spec],
      intents: ["report", "diagnosis"],
      priority: 100,
    });
  }
  suggestions.sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));
  return suggestions;
}

export function renderTemplateSelectionYAML(policy) {
  const version = policy.version || 1;
  let out = `version: ${version}\n\ntemplates:\n`;
  for (const rule of policy.templates || []) {
    out += `  - id: ${rule.id}\n`;
    out += `    playbook: ${rule.playbook}\n`;
    out += `    template: ${rule.template}\n`;
    out += `    type: ${rule.type}\n`;
    if (rule.domain) out += `    domain: ${rule.domain}\n`;
    out += writeSimpleYAMLArray("covers", rule.covers);
    out += writeSimpleYAMLArray("intents", rule.intents);
    out += `    priority: ${rule.priority || 0}\n`;
  }
  return out;
}

function writeSimpleYAMLArray(key, values) {
  if (!values?.length) return "";
  let out = `    ${key}:\n`;
  for (const value of values) out += `      - ${value}\n`;
  return out;
}

export function templateDoctorJSON(result) {
  const out = {
    status: result.status,
    selectionPath: result.selectionPath,
    suggestionWritten: Boolean(result.suggestionWritten),
  };
  if (result.rules?.length) out.rules = result.rules.map(templateSelectionRuleJSON);
  if (result.errors?.length) out.errors = result.errors;
  if (result.warnings?.length) out.warnings = result.warnings;
  if (result.suggestions?.length) out.suggestions = result.suggestions.map(templateSelectionRuleJSON);
  if (result.suggestionPath) out.suggestionPath = result.suggestionPath;
  return out;
}

function templateSelectionRuleJSON(rule) {
  const out = {
    id: rule.id,
    playbook: rule.playbook,
    template: rule.template,
    type: rule.type,
  };
  if (rule.domain) out.domain = rule.domain;
  if (rule.covers?.length) out.covers = rule.covers;
  if (rule.intents?.length) out.intents = rule.intents;
  if (rule.priority) out.priority = rule.priority;
  return out;
}

function fileExists(filePath) {
  return isRegularFile(filePath);
}

function pathBase(logical) {
  const parts = String(logical || "").split("/");
  if (parts.length === 0) return logical;
  return parts[parts.length - 1];
}

export function stableTemplateSelectionID(templatePath) {
  let id = String(templatePath || "");
  if (id.startsWith("templates/")) id = id.slice("templates/".length);
  if (id.endsWith(".md")) id = id.slice(0, -".md".length);
  if (id.startsWith("reports/")) id = id.slice("reports/".length);
  return id.replaceAll("/", "_").replaceAll("-", "_");
}

export function isReportTemplatePath(logical) {
  return String(logical || "").startsWith("reports/") && pathBase(logical) === "template.md";
}

export function isAllowedTemplateSelectionPath(logical) {
  return String(logical || "").startsWith("templates/") || isReportTemplatePath(logical);
}
